telethon/telegram_client.py

```diff
@@ -239,25 +232,52 @@ def start(self, phone=None, bot_token=None, force_sms=False, code_callback=None,
         if self.is_user_authorized():
             return self
 
+        two_step_detected = False
+
         if bot_token:
             self(ImportBotAuthorizationRequest(
                 flags=0, bot_auth_token=bot_token,
                 api_id=self.api_id, api_hash=self.api_hash
             ))
         elif phone or self._phone:
             phone = phone or self._phone
-            self.send_code_request(phone)
-            provided_code = str(code_callback())
+
+            self.send_code_request(phone, force_sms)
+
+            attempts = 0
+            while attempts < 3:
+                try:
+                    # Normal login
+                    provided_code = str(code_callback())
+
+                    # raises SessionPasswordNeededError if 2FA enabled
+                    signed_in = bool(self.sign_in(phone, provided_code))
+                    if signed_in:
+                        print("Signed in successfully as {}".format(self.get_me()))
```

Comment:
`.get_me()` is a bit silly. `.sign_in` already returns the user, that's an unnecessary extra call…